### ISSUE TITLE
Test failure on win32 (FilePathTestCase.testGetAndSet)

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#11195]
|[<img alt="teratorn's avatar" src="https://avatars.githubusercontent.com/u/154069?s=50" width="50" height="50">](https://github.com/teratorn)| @teratorn reported|
|-|-|
|Trac ID|trac#1705|
|Type|defect|
|Created|2006-05-06 09:52:03Z|

twisted.test.test_paths.FilePathTestCase.testGetAndSet fails on win32 due to the inconsistent behavior of os.rename across platforms, and is currently marked TODO.

On UNIX os.rename will overwrite the destination in most cases. On win32 it will never overwrite.

FilePath's setContent method relies on overwriting the path with os.rename if it already exists.

Why, exactly, does setContent create a tmp file and move it over the destination instead of just writing the content directly to the destination?

<details><summary>Searchable metadata</summary>

```
trac-id__1705 1705
type__defect defect
reporter__teratorn teratorn
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__win32 win32
time__1146909123000000 1146909123000000
changetime__1147654179000000 1147654179000000
version__2_0 2.0
owner__teratorn teratorn
cc__glyph cc__exarkun
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="teratorn's avatar" src="https://avatars.githubusercontent.com/u/154069?s=50" width="50" height="50">](https://github.com/teratorn)|@teratorn set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="teratorn's avatar" src="https://avatars.githubusercontent.com/u/154069?s=50" width="50" height="50">](https://github.com/teratorn)|@teratorn commented|
|-|-|

I guess the reason is that os.rename is atomic. If anything bad happens during the setContent() call, you know that the original file is still intact.

So, do we really care about that behavior? If so, I can look in to how to implement it on Windows. Otherwise, we can delete the original file before the rename, if it exists. Or just open the destination and write to it directly.

Glyph, Moshez, I'm CC'ing you because svn blame says you are the likely people to bug :)
Comments?

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph commented|
|-|-|

You've correctly identified the reason why, and yes, we care about that behavior.  It is particularly handy on UNIX if, for example, you have write permission to the directory but not the file in question.

It may be impossible to provide this particular guarantee on Windows though, without some kind of atomic rename.  In that case I'm not sure what the right thing to do is.

**Comment by twisted-trac** [comment]
|[<img alt="teratorn's avatar" src="https://avatars.githubusercontent.com/u/154069?s=50" width="50" height="50">](https://github.com/teratorn)|@teratorn commented|
|-|-|

Well, I've Googled, I've scoured the PSDK, and I've searched the MSDN all pretty hard. Found quite a few people who claim an atomic rename isn't possible on win32, and nothing to the contrary :(

We could check if the platform is win32, and only then write the data directly to the destination?

**Comment by twisted-trac** [comment]
|[<img alt="teratorn's avatar" src="https://avatars.githubusercontent.com/u/154069?s=50" width="50" height="50">](https://github.com/teratorn)|@teratorn commented|
|-|-|

One possibility is MoveFileEx() which is available on NT4 and above. It allows a flag that will cause the destination to be overwritten if it exists.

I'm still not sure if it's atomic though. It doesn't say one way or the other in the official documentation, though I've seen a few mailing list posts that claim it is. Who knows?

**Comment by twisted-trac** [comment]
|[<img alt="teratorn's avatar" src="https://avatars.githubusercontent.com/u/154069?s=50" width="50" height="50">](https://github.com/teratorn)|@teratorn set owner to @glyph<br>@teratorn set status to `new`|
|-|-|

After further investigation, I'm satisfied that an atomic rename() on win32 isn't possible using a any standard APIs.

Branch is ready for review. Current implementation writes content directly to the destination file if the platform is Windows.

Glyph, I'm assigning to you. Please reassign or let me know if you want someone else to look at it.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set status to `assigned`|
|-|-|

I will review this later tonight.  Thanks for investigating.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

I think setContent should still write to a separate file on Windows.  It can delete the original and then safely rename the new file.  This is better than writing directly to the new file since with a little more application-level code it is at least possible to recover from a mid-setContent failure using the separate-file scheme, but completely impossible using the direct-write scheme.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set owner to @teratorn<br>@glyph set status to `new`|
|-|-|

I agree with what JP said; it should still do write-aside.  As the size of the value value provided to setContent increases, the probability of dying in mid-write increases whereas the probability of dying between the delete and the rename remains constant.  (Or, you know, constant to within one unit of "NTFS sucks ass")

**Comment by twisted-trac** [comment]
|[<img alt="teratorn's avatar" src="https://avatars.githubusercontent.com/u/154069?s=50" width="50" height="50">](https://github.com/teratorn)|@teratorn set owner to @glyph|
|-|-|

Yep makes sense. Changes committed. OK to merge?

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set owner to @teratorn|
|-|-|

Yeah.  Merge away.

**Comment by twisted-trac** [comment]
|[<img alt="teratorn's avatar" src="https://avatars.githubusercontent.com/u/154069?s=50" width="50" height="50">](https://github.com/teratorn)|@teratorn set status to `closed`|
|-|-|

(In [16818]) Merge testGetAndSet-win32-1705

Author: teratorn 
Review: glyph 
Fixes [#11195](https://github.com/twisted/twisted/issues/11195)

On win32, make FilePath.setContent() remove the destination file first because os.rename() will not overwrite an existing file. There doesn't appear to be any way to do atomic renames on win32 so this is as safe as we can make it.
